Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSPatcher.mm

```diff
@@ -0,0 +1,160 @@
+
+#include "pfPatcher/pfPatcher.h"
+#include "pfPatcher/plManifests.h"
+#include "plFileSystem.h"
+#include "plNetGameLib/plNetGameLib.h"
+
+class Patcher
+{
+   public:
+    PLSPatcher* parent;
+    void IOnPatchComplete(ENetError result, const ST::string& msg);
+    void IOnProgressTick(uint64_t curBytes, uint64_t totalBytes, const ST::string& status);
+    void IOnDownloadBegin(const plFileName& file);
+};
+
+@interface PLSPatcher ()
+@property BOOL selfPatched;
+@property pfPatcher* patcher;
+@property NSTimer* networkPumpTimer;
+@property Patcher cppPatcher;
+@end
+
+@implementation PLSPatcher
+
+- (id)init
+{
+    self = [super init];
+    self.selfPatched = false;
+
+    _cppPatcher.parent = self;
+
+    self.patcher = new pfPatcher();
+    _patcher->OnFileDownloadBegin(
+        std::bind(&Patcher::IOnDownloadBegin, _cppPatcher, std::placeholders::_1));
+    _patcher->OnProgressTick(std::bind(&Patcher::IOnProgressTick, _cppPatcher,
+                                       std::placeholders::_1, std::placeholders::_2,
+                                       std::placeholders::_3));
+    _patcher->OnCompletion(std::bind(&Patcher::IOnPatchComplete, _cppPatcher, std::placeholders::_1,
+                                     std::placeholders::_2));
+    _patcher->OnFileDownloadDesired(IApproveDownload);
+    _patcher->OnSelfPatch([&](const plFileName& file) {});
+    _patcher->OnRedistUpdate([&](const plFileName& file) {});
+
+    self.networkPumpTimer = [NSTimer timerWithTimeInterval:1.0 / 1000.0
+                                                   repeats:true
+                                                     block:^(NSTimer* _Nonnull timer) {
+                                                         hsTimer::IncSysSeconds();
+                                                         NetClientUpdate();
+                                                     }];
+
+    return self;
+}
+
+- (void)start
+{
+    [[NSRunLoop mainRunLoop] addTimer:self.networkPumpTimer forMode:NSDefaultRunLoopMode];
+    self.patcher->RequestManifest(plManifest::ClientManifest());
+    self.patcher->Start();
+}
+
+void Patcher::IOnDownloadBegin(const plFileName& file)
+{
+    NSString* fileName = [NSString stringWithSTString:file.AsString()];
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [parent.delegate patcher:parent beganDownloadOfFile:fileName];
+    });
+}
+
+void Patcher::IOnProgressTick(uint64_t curBytes, uint64_t totalBytes, const ST::string& status)
+{
+    NSString* statusString = [NSString stringWithSTString:status];
+
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [parent.delegate patcher:parent
+                 updatedProgress:statusString
+                       withBytes:curBytes
+                           outOf:totalBytes];
+    });
+}
+
+bool IApproveDownload(const plFileName& file)
+{
+    ST::string ext = file.GetFileExt();
+    // nothing from Windows, please
+    return ext != "exe" && ext != "pdb" && ext != "dll";
```

Comment:
It might be better to have an `unordered_set` of extensions you don't want and test against that. I realize that this is ObjC land, so maybe I don't understand the parameters here, but I'm thinking something like....

```C++
static std::unordered_set<ST::string, ST::hash_i, ST::equal_i> extBlacklist{
    "exe"_st,
    "dll"_st,
    "pdb"_st
};
return extBlacklist.find(file.GetFileExt()) == extBlacklist.end();
```

I have a thing against long `if`s ;)

Reply:
I'll take a look. This is also a temporary chunk. When the Mac can patch from it's own manifest this code will go away.